### ISSUE TITLE
initial file open sometimes replaces character under cursor with 'g'

### ISSUE THREAD

**Issue by rcaputo** [neovim/neovim#29047]
### Problem

Frequently, but not always, the first file opened from the command line will have the character under the cursor replaced by the letter 'g'. This seems to only occur when `set nottimeout` is enabled. The inconsistency suggests a startup race condition.

### Steps to reproduce

`nvim --clean --cmd 'set nocompatible' --cmd 'set nottimeout' ~/.vimrc`

Removing the `--cmd 'set nottimeout'` seems to greatly reduce if not eliminate the issue.

### Expected behavior

I expect the file to appear in the buffer as it does on disk. The buffer state should initially be unmodified.

### Neovim version (nvim -v)

% nvim -v NVIM v0.10.0 Build type: Release LuaJIT 2.1.1713773202

### Vim (not Nvim) behaves the same?

no, vim 9.0.1544 (nor a previous forgotten nvim 0.9.x version)

### Operating system/version

macOS 11.7.10

### Terminal name/version

Reproduced with two: iTerm2, Wezterm

### $TERM environment variable

iTerm2:xterm-256color, Wezterm:wezterm

### Installation

brew install neovim

**Comment by justinmk** [comment]
Does it actually change the buffer contents or is it only the display? Does CTRL-L clear it?

**Comment by rcaputo** [comment]
Changes the buffer contents. Puts it into a modified state. Attempting to immediately `:q` will result in
```
E37: No write since last change
E162: No write since last change for buffer "/Users/me/.vimrc"
```

**Comment by nyorem** [comment]
I'm having the same problem: I tend to use Ctrl-Z+fg to switch between neovim and my terminal and almost everytime I come back to neovim, some random text is being pasted.

On my side, I had "set ttimeoutlen=0" in my configuration. Removing it fixes the issue.

**Comment by zeertzjq** [comment]
It seems that the behavior of 'nottimeout' is not documented, and its current behavior is the opposite of Vim's: in Vim 'nottimeout' means infinite timeout, while in Nvim 'nottimeout' means zero timeout.

The following change seems to make 'nottimeout' behave like infinite timeout as in Vim:
```diff
diff --git a/src/nvim/tui/input.c b/src/nvim/tui/input.c
index f1594dfcb..b7fbcee74 100644
--- a/src/nvim/tui/input.c
+++ b/src/nvim/tui/input.c
@@ -463,7 +463,7 @@ static void tk_getkeys(TermInput *input, bool force)
     uv_timer_stop(&input->timer_handle);
     uv_timer_start(&input->timer_handle, tinput_timer_cb, (uint64_t)input->ttimeoutlen, 0);
   } else {
-    tk_getkeys(input, true);
+    tk_getkeys(input, false);
   }
 }
 
```

**Comment by rcaputo** [comment]
I looked to see how `p_timeout` is treated elsewhere, in case it has a similar interpretation when false.

I found two instances. Both involve `p_ttimeout`.

Here ESC in insert mode seems to always be treated as a single keystroke if both `p_timeout` and `p_ttimeout` are false: https://github.com/neovim/neovim/blob/5c33815448e11b514678f39cecc74e68131d4628/src/nvim/getchar.c#L2637-L2638

But a simple test passes: `set notimeout` and `set nottimeout`, go into insert mode, and press PageUp or PageDown. This works fine, so the code isn't doing what I thought it did.

The wait time in the `inchar()` call above is hard-coded at 25 milliseconds. Should it be `(int)p_ttm`?

Here the semantics seem to be correct. If both `p_timeout` and `p_ttimeout` are false, then the wait time for the next keystroke may be indefinite: https://github.com/neovim/neovim/blob/5c33815448e11b514678f39cecc74e68131d4628/src/nvim/getchar.c#L2827-L2841

**Comment by drewfish** [comment]
I ran into this too when I had the following set
```
set nottimeout
set timeout
set timeoutlen=1000
set ttimeoutlen=-1
```

I had to comment out BOTH `nottimeout` and `ttimeoutlen` for the issue to go away.

When I had the issue and the following set, the opening a file would lead to the first character of the line from last edit being changed to a `g` AND the cursor would be at the start of the last word on the previous line.
```
autocmd BufReadPost * if line("'\"") > 0 && line("'\"") <= line("$") | exe "normal g'\"" | endif
```

**Comment by 3rd** [comment]
The issue is back on NVIM v0.11.0-nightly+59cbe64, causing nasty side effects like clipboard interaction when ttimeout=false.
A previous v0.11.0-dev version works fine, spend quite a bit debugging config and plugins, and turns out some plugin also sets nottimeout=false which is going to be problematic if that breaks things when opening Neovim.

**Comment by superobertking** [comment]
Ran into the same issue recently. It seems the combined keystrokes are "rgb", which is likely related to terminal reporting.

**Comment by msteed** [comment]
I can reproduce similar buffer corruption on startup with this `init.vim`:

```
set ttimeoutlen=10
colorscheme darkblue
```

The problem may be reproducible without calling `colorscheme`, but having that present causes lots of buffer corruption about 30% of the time that I start nvim. Instead of a single character being replaced, I see tens or hundreds of characters written to the buffer. Obviously there are many vim commands running as part of `colorscheme`, which complicates things a bit, but it does seem to make the issue easier to reproduce.

I see this with nvim 0.10.3 and 0.10.2. Vim 9.1 is unaffected.

**Comment by 3rd** [comment]
Still getting this (why was my last comment deleted?) with ttimeout=false, recent versions replace the buffer contents with "5c" on open and opening Neovim with ttimeout=true fixes this. It doesn't happen with an empty config, and I'm not sure what triggers it.

**Comment by justinmk** [comment]
> why was my last comment deleted?

is your comment adding new information to this issue?

**Comment by 3rd** [comment]
> > why was my last comment deleted?
> 
> is your comment adding new information to this issue?

I don't remember, it was deleted not hidden. It was adding at least that it wasn't happening in some cases until recently.
neovim/neovim/e4bc8b5967d22840c1e52c97acab0f77107cd48c (2024-12-29) didn't have the issue

**Comment by justinmk** [comment]
Yes, and your latest comment is the same. We already have plenty of confirmations, and don't need more. That is why your comment was deleted. "It started recently" doesn't give any useful information, especially since the details of your system "before recently" are unknown. 

Bisecting to an exact commit would be useful.

**Comment by 3rd** [comment]
> Yes, and your latest comment is the same. We already have plenty of confirmations, and don't need more. That is why your comment was deleted. "It started recently" doesn't give any useful information, especially since the details of your system "before recently" are unknown.
> 
> Bisecting to an exact commit would be useful.

Ok, but the one you deleted now had a quick to paste workaround for anyone who hits the same issue.
Whatever, you know your priorities, I find it weird to delete comments in general.

**Comment by clason** [comment]
GitHub will start hiding comments indiscriminately if there are too many (including hidden ones). So "empty" comments are actively detrimental.

**Comment by iniw** [comment]
To anyone that runs into this: set `ttimeoutlen=0` instead of setting `nottimeout`. It achieves the same thing (or at least seems to) while not suffering from this peculiar issue.
